corehq/ex-submodules/casexml/apps/phone/restore.py

```diff
@@ -183,9 +215,46 @@ def compose(self):
             body=self.response_body.getvalue(),
             end=self.closing_tag
         )
+        self.finalized = True
+        self.close()
 
-    def __str__(self):
-        return self.compose()
+    def get_cache_payload(self, full=False):
+        return {
+            'is_file': False,
+            'data': self.response
+        }
+
+    def as_string(self):
+        return self.response
+
+    def get_http_response(self):
+        return HttpResponse(self.response, mimetype="text/xml")
+
+
+class CachedResponse(object):
+    def __init__(self, payload):
+        if isinstance(payload, basestring):
```

Comment:
This is coming from the cache. If it comes from the synclog then it will be a string (the attachment contents) but if it comes from redis it will be a dict.